pandas/core/arrays/boolean.py

```diff
@@ -576,17 +581,38 @@ def logical_method(self, other):
 
             # numpy will show a DeprecationWarning on invalid elementwise
             # comparisons, this will raise in the future
-            with warnings.catch_warnings():
-                warnings.filterwarnings("ignore", "elementwise", FutureWarning)
-                with np.errstate(all="ignore"):
-                    result = op(self._data, other)
+            if lib.is_scalar(other) and np.isnan(
+                other
+            ):  # TODO(NA): change to libmissing.NA:
+                result = self._data
+                mask = True
+            else:
+                with warnings.catch_warnings():
+                    warnings.filterwarnings("ignore", "elementwise", FutureWarning)
+                    with np.errstate(all="ignore"):
+                        result = op(self._data, other)
 
             # nans propagate
             if mask is None:
                 mask = self._mask
             else:
                 mask = self._mask | mask
 
+            # Kleene-logic adjustments to the mask.
+            if op.__name__ in {"or_", "ror_"}:
+                mask[result] = False
+            elif op.__name__ in {"and_", "rand_"}:
+                mask[~self._data & ~self._mask] = False
+                if other_is_booleanarray:
+                    mask[~other & ~omask] = False
+                elif lib.is_scalar(other) and np.isnan(other):  # TODO(NA): change to NA
+                    mask[:] = True
+                # Do we ever assume that masked values are False?
```

Comment:
@jorisvandenbossche do we assume / promise that `self._data[self._mask]` is all False?

Reply:
No, there is currently no guarantee on the value in `_data` that is masked